classes/local/execution/engine.php

```diff
@@ -535,14 +535,21 @@ public function abort(?\Throwable $reason = null) {
         }
 
         $this->set_current_step(null);
-        $this->log('Engine: aborting steps');
+        $this->log('Engine: aborting steps', $message ? ['reason' => $message] : [], Logger::NOTICE);
         $this->exception = $reason;
         foreach ($this->enginesteps as $enginestep) {
-            $this->set_current_step($enginestep);
-            $enginestep->abort();
+            $status = $enginestep->status;
+            if ($status !== self::STATUS_FINISHED && !in_array($status, self::STATUS_TERMINATORS)) {
```

Comment:
This check is used a fair bit, might be worth extracting to an is_terminated static method. Might not be worth, but might be a bit more readable/writeable

Reply:
I had thought about that also. It's currently used in 2 places, within the same main function. The definition of a terminator seems to be different than a step which is in a finished state, which is why the finished was added as a separate check, as opposed to adding finished to the list of status terminators. One example I can think of is that steps in a finished state, will not "release" the resource until they transition to finalised (e.g. on_finalise)

In terms of keeping things pretty DRY, I'm all for it, though I prefer approaching it with more pragmatism and avoid abstracting too early. I personally follow the rule of 3.
https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)

Reply:
Perhaps an alternative would be to replace the check with something more readable such as this:
```suggestion
            $isdone = status !== self::STATUS_FINISHED && !in_array($status, self::STATUS_TERMINATORS);
            if ($isdone) {
```

Reply:
Yeah this is a nice middle ground.